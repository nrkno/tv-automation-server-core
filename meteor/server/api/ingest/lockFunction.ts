import { Meteor } from 'meteor/meteor'
import _ from 'underscore'
import { PartId } from '../../../lib/collections/Parts'
import { DBRundown } from '../../../lib/collections/Rundowns'
import { SegmentId } from '../../../lib/collections/Segments'
import { ShowStyleCompound } from '../../../lib/collections/ShowStyleVariants'
import { StudioId } from '../../../lib/collections/Studios'
import { clone, protectString } from '../../../lib/lib'
import { ReadOnlyCache } from '../../cache/CacheBase'
import { pushWorkToQueue } from '../../codeControl'
import { WrappedShowStyleBlueprint } from '../blueprints/cache'
import { getRundownsSegmentsAndPartsFromCache } from '../playout/lib'
import { PlayoutLockFunctionPriority, runPlayoutOperationWithLock } from '../playout/lockFunction'
import { profiler } from '../profiler'
import { CacheForIngest } from './cache'
import { BeforePartMap, CommitIngestOperation } from './commit'
import { LocalIngestRundown, RundownIngestDataCache } from './ingestCache'
import { getRundown, getRundownId } from './lib'

export interface CommitIngestData {
	/** Segment Ids which had any changes */
	changedSegmentIds: SegmentId[]
	/** Segments to be removed or orphaned */
	removedSegmentIds: SegmentId[]
	/**
	 * Segments that had their ids changed. This helps then be orphaned in the correct place
	 * eg, whole segment is renamed and middle part deleted
	 * Note: Only supported for MOS, not 'normal' ingest operations
	 */
	renamedSegments: Map<SegmentId, SegmentId>

	/** Whether the rundown should be removed or orphaned */
	removeRundown: boolean

	/** ShowStyle, if loaded to reuse */
	showStyle: ShowStyleCompound | undefined
	/** Blueprint, if loaded to reuse */
	blueprint: WrappedShowStyleBlueprint | undefined
}

export enum UpdateIngestRundownAction {
	DELETE = 'delete',
}

/**
 * Perform an ingest update operation on a rundown
 * This will automatically do some post-update data changes, to ensure the playout side (partinstances etc) is updated with the changes
 * @param context Contextual information for the call to this function. to aid debugging
 * @param studioId Id of the studio the rundown belongs to
 * @param rundownExternalId ExternalId of the rundown to lock
 * @param updateCacheFcn Function to mutate the ingestData. Throw if the requested change is not valid. Return undefined to indicate the ingestData should be deleted
 * @param calcFcn Function to run to update the Rundown. Return the blob of data about the change to help the post-update perform its duties. Return null to indicate that nothing changed
 */
// runPlayoutOperationWithCacheFromStudioOperation
export async function runIngestOperationWithCache(
	context: string,
	studioId: StudioId,
	rundownExternalId: string,
	updateCacheFcn: (
		oldIngestRundown: LocalIngestRundown | undefined
	) => LocalIngestRundown | UpdateIngestRundownAction,
	calcFcn: (
		cache: CacheForIngest,
		newIngestRundown: LocalIngestRundown | undefined,
		oldIngestRundown: LocalIngestRundown | undefined
	) => Promise<CommitIngestData | null>
): Promise<void> {
	return ingestLockFunctionInner(context, rundownExternalId, async () => {
		const span = profiler.startSpan(`ingestLockFunction.${context}`)

		// Load the old ingest data
		const rundownId = getRundownId(studioId, rundownExternalId)
		const pIngestCache = CacheForIngest.create(studioId, rundownExternalId)
		const ingestObjCache = await RundownIngestDataCache.create(rundownId)

		// Recalculate the ingest data
		const oldIngestRundown = ingestObjCache.fetchRundown()
		const updatedIngestRundown = updateCacheFcn(clone(oldIngestRundown))
		let newIngestRundown: LocalIngestRundown | undefined
		switch (updatedIngestRundown) {
			// case UpdateIngestRundownAction.REJECT:
			// 	// Reject change
			// 	return
			case UpdateIngestRundownAction.DELETE:
				ingestObjCache.delete()
				newIngestRundown = undefined
				break
			default:
				ingestObjCache.update(updatedIngestRundown)
				newIngestRundown = updatedIngestRundown
				break
		}
		// Start saving the ingest data
		const pSaveIngestChanges = ingestObjCache.saveToDatabase()

		try {
			const ingestCache = await pIngestCache

			// Load any 'before' data for the commit
			const beforeRundown = ingestCache.Rundown.doc
			const beforePartMap = generatePartMap(ingestCache)

			const commitData = await calcFcn(ingestCache, newIngestRundown, oldIngestRundown)
			if (commitData) {
				// The change is accepted. Perform some playout calculations and save it all
				await CommitIngestOperation(ingestCache, beforeRundown, beforePartMap, commitData)
			} else {
				// Should be no changes
				ingestCache.assertNoChanges()
			}
		} finally {
			// Ensure we save the ingest data
			await pSaveIngestChanges

			span?.end()
		}
	})
}

/**
 * Perform an operation on a rundown that has no effect on playout
 * Warning: this will not attempt to update playout or perform the usual ingest post-changes. Be very careful what is changed as it could cause weird behaviour
 * @param context Contextual information for the call to this function. to aid debugging
 * @param studioId Id of the studio the rundown belongs to
 * @param rundownExternalId ExternalId of the rundown to lock
 * @param fcn Function to run while holding the lock
 * @param playlistLock If the playlist lock is already held, supply it here to avoid trying to reaquire the lock
 */
export async function runIngestOperationFromRundown(
	context: string,
	refRundown: DBRundown,
	fcn: (ingestCache: CacheForIngest) => Promise<void>
): Promise<void> {
	const refRundownId = refRundown._id
	const studioId = refRundown.studioId
	const rundownExternalId = refRundown.externalId

	const targetRundownId = getRundownId(studioId, rundownExternalId)
	if (targetRundownId !== refRundownId) {
		throw new Meteor.Error(
			500,
			`Rundown id cannot run ingest operations, as it's _id "${refRundownId}" does not match the expected "${targetRundownId}". Perhaps it was restored from a snapshot?`
		)
	}

	return ingestLockFunctionInner(context, rundownExternalId, async () => {
		const ingestCache = await CacheForIngest.create(studioId, rundownExternalId)
		const beforeRundown = getRundown(ingestCache)

		// Do the function
		await fcn(ingestCache)

		// Check the change isn't making a mess
		const afterRundown = getRundown(ingestCache)
		if (beforeRundown.playlistId !== afterRundown.playlistId)
			throw new Meteor.Error(
				500,
				`RundownPlaylist id for Rundown "${ingestCache.RundownId}" cannot be changed during a ingestRundownOnlyLockFunction`
			)

		await runPlayoutOperationWithLock(
			null,
			context,
			beforeRundown.playlistId,
			PlayoutLockFunctionPriority.MISC,
			async () => {
				// This needs to be inside the playout lock to ensure that a take doesnt happen mid update
				await ingestCache.saveAllToDatabase()
			}
		)
	})
}

async function ingestLockFunctionInner(
	context: string,
	rundownExternalId: string,
	fcn: () => Promise<void>
): Promise<void> {
	return pushWorkToQueue(`rundown_ingest_${rundownExternalId}`, context, fcn)
}

function generatePartMap(cache: ReadOnlyCache<CacheForIngest>): BeforePartMap {
	const rundown = cache.Rundown.doc
	if (!rundown) return new Map()

	const segmentsAndParts = getRundownsSegmentsAndPartsFromCache(cache.Parts, cache.Segments, [rundown])
	const existingRundownParts = _.groupBy(segmentsAndParts.parts, (part) => part.segmentId)

	const res = new Map<SegmentId, Array<{ id: PartId; rank: number }>>()
	for (const [segmentId, parts] of Object.entries(existingRundownParts)) {
		res.set(
			protectString(segmentId),
			parts.map((p) => ({ id: p._id, rank: p._rank }))
		)
	}
	return res
}
