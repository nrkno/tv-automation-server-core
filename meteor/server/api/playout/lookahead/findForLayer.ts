import { PartInstanceId } from '../../../../lib/collections/PartInstances'
import { Part } from '../../../../lib/collections/Parts'
import { OnGenerateTimelineObjExt, TimelineObjRundown } from '../../../../lib/collections/Timeline'
import { profiler } from '../../profiler'
import { sortPieceInstancesByStart } from '../pieces'
import { findLookaheadObjectsForPart } from './findObjects'
import { PartAndPieces, PartInstanceAndPieceInstances } from './util'

export interface LookaheadResult {
	timed: Array<TimelineObjRundown & OnGenerateTimelineObjExt>
	future: Array<TimelineObjRundown & OnGenerateTimelineObjExt>
}

export function findLookaheadForLayer(
	currentPartInstanceId: PartInstanceId | null,
	partInstancesInfo: PartInstanceAndPieceInstances[],
	previousPartInstanceInfo: PartInstanceAndPieceInstances | undefined,
	orderedPartInfos: Array<PartAndPieces>,
	layer: string,
	lookaheadTargetFutureObjects: number,
	lookaheadMaxSearchDistance: number
): LookaheadResult {
	const span = profiler.startSpan(`findLookaheadForlayer.${layer}`)
	const res: LookaheadResult = {
		timed: [],
		future: [],
	}

	// Track the previous info for checking how the timeline will be built
	let previousPart: Part | undefined
	if (previousPartInstanceInfo) {
		previousPart = previousPartInstanceInfo.part.part
	}

	// Generate timed/future objects for the partInstances
	for (const partInstanceInfo of partInstancesInfo) {
		if (!partInstanceInfo.onTimeline && lookaheadMaxSearchDistance <= 0) break

		const partInfo: PartAndPieces = {
			part: partInstanceInfo.part.part,
			pieces: sortPieceInstancesByStart(partInstanceInfo.allPieces, partInstanceInfo.nowInPart),
		}

		const objs = findLookaheadObjectsForPart(
			currentPartInstanceId,
			layer,
			previousPart,
			partInfo,
			partInstanceInfo.part._id
		)

		if (partInstanceInfo.onTimeline) {
			res.timed.push(...objs)
		} else {
			res.future.push(...objs)
		}

		previousPart = partInfo.part
	}

	if (lookaheadMaxSearchDistance > 1 && lookaheadTargetFutureObjects > 0) {
		for (const partInfo of orderedPartInfos.slice(0, lookaheadMaxSearchDistance - 1)) {
			// Stop if we have enough objects already
			if (res.future.length >= lookaheadTargetFutureObjects) {
				break
			}

			if (partInfo.pieces.length > 0 && partInfo.part.isPlayable()) {
				const objs = findLookaheadObjectsForPart(currentPartInstanceId, layer, previousPart, partInfo, null)
				res.future.push(...objs)
				previousPart = partInfo.part
			}
		}
	}

	if (span) span.end()
	return res
}
