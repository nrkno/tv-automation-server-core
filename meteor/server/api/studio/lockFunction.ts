import { StudioId } from '../../../lib/collections/Studios'
import { ReadOnlyCache } from '../../cache/CacheBase'
import { pushWorkToQueue } from '../../codeControl'
import { profiler } from '../profiler'
import { CacheForStudio } from './cache'

/** Priority for handling of synchronous events. Higher value means higher priority */
export enum StudioLockFunctionPriority {
	MISC = 0,
	/** Events initiated from user, for playout */
	USER_PLAYOUT = 10,
	/** Events initiated from playout-gateway callbacks */
	CALLBACK_PLAYOUT = 20,
}

export interface StudioLock {
	readonly _studioId: StudioId
}

export function isStudioLock(obj: any): obj is StudioLock {
	const obj0 = obj as StudioLock
	return !!obj0._studioId
}

export function getStudioIdFromCacheOrLock(
	cacheOrLock: StudioLock | Pick<ReadOnlyCache<CacheForStudio>, 'Studio'>
): StudioId {
	if (isStudioLock(cacheOrLock)) {
		return cacheOrLock._studioId
	} else {
		return cacheOrLock.Studio.doc._id
	}
}

/**
 * Lock the studio and load a cache of data about the studio
 * @param context Contextual information for the call to this function. to aid debugging
 * @param studioId Id of the studio to lock
 * @param fcn Function to run while holding the lock
 */
export async function runStudioOperationWithCache<T>(
	context: string,
	studioId: StudioId,
	priority: StudioLockFunctionPriority,
	fcn: (cache: CacheForStudio) => Promise<T>
): Promise<T> {
	return runStudioOperationWithLock(context, studioId, priority, async () => {
		const cache = await CacheForStudio.create(studioId)

		const res = await fcn(cache)

		await cache.saveAllToDatabase()

		return res
	})
}

/**
 * Lock the studio
 * @param context Contextual information for the call to this function. to aid debugging
 * @param studioId Id of the studio to lock
 * @param fcn Function to run while holding the lock
 */
export async function runStudioOperationWithLock<T>(
	context: string,
	studioId: StudioId,
	priority: StudioLockFunctionPriority,
	fcn: (lock: StudioLock) => Promise<T>
): Promise<T> {
	return pushWorkToQueue(
		`studio_${studioId}`,
		context,
		async () => {
			const span = profiler.startSpan(`studioLockFunction.${context}`)
			const res = await fcn({ _studioId: studioId })
			span?.end()
			return res
		},
		priority
	)
}
