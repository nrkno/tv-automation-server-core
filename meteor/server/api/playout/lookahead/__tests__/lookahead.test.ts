import '../../../../../__mocks__/_extendJest'
import {
	setupDefaultStudioEnvironment,
	DefaultEnvironment,
	setupDefaultRundownPlaylist,
} from '../../../../../__mocks__/helpers/database'
import { DBRundown, RundownId, Rundowns } from '../../../../../lib/collections/Rundowns'
import { RundownPlaylist, RundownPlaylistId, RundownPlaylists } from '../../../../../lib/collections/RundownPlaylists'
import { getCurrentTime, getRandomId, protectString } from '../../../../../lib/lib'
import { SegmentId } from '../../../../../lib/collections/Segments'
import { DBPart, Part, PartId, Parts } from '../../../../../lib/collections/Parts'
import { LookaheadMode, PlaylistTimingType, TSR } from '@sofie-automation/blueprints-integration'
import { MappingsExt, Studios } from '../../../../../lib/collections/Studios'
import { OnGenerateTimelineObjExt, TimelineObjRundown } from '../../../../../lib/collections/Timeline'
import { PartAndPieces, PartInstanceAndPieceInstances } from '../util'
import { getLookeaheadObjects } from '..'
import { SelectedPartInstancesTimelineInfo } from '../../timeline'
import { testInFiber } from '../../../../../__mocks__/helpers/jest'
import { PlayoutLockFunctionPriority, runPlayoutOperationWithCache } from '../../lockFunction'
import { LOOKAHEAD_DEFAULT_SEARCH_DISTANCE } from '../../../../../lib/constants'

jest.mock('../findForLayer')
type TfindLookaheadForLayer = jest.MockedFunction<typeof findLookaheadForLayer>
const findLookaheadForLayerMock = findLookaheadForLayer as TfindLookaheadForLayer
import { findLookaheadForLayer } from '../findForLayer'

jest.mock('../util')
type TgetOrderedPartsAfterPlayhead = jest.MockedFunction<typeof getOrderedPartsAfterPlayhead>
const getOrderedPartsAfterPlayheadMock = getOrderedPartsAfterPlayhead as TgetOrderedPartsAfterPlayhead
import { getOrderedPartsAfterPlayhead } from '../util'

describe('Lookahead', () => {
	let env: DefaultEnvironment
	let playlistId: RundownPlaylistId
	// let rundownId: RundownId
	// let segmentId: SegmentId
	let partIds: PartId[]

	beforeEach(async () => {
		findLookaheadForLayerMock.mockReset().mockReturnValue({ timed: [], future: [] }) // Default mock
		getOrderedPartsAfterPlayheadMock.mockReset().mockReturnValue([])

		env = await setupDefaultStudioEnvironment()
		const mappings: MappingsExt = {}
		for (const k of Object.keys(LookaheadMode)) {
			if (isNaN(parseInt(k))) {
				mappings[k] = {
					device: TSR.DeviceType.ABSTRACT,
					deviceId: protectString('fake0'),
					lookahead: LookaheadMode[k],
					// lookaheadDepth: 0,
					// lookaheadMaxSearchDistance: 0,
				}
			}
		}
		Studios.update(env.studio._id, { $set: { mappings } })
		env.studio.mappings = mappings
		;({ playlistId } = setupDefaultRundownPlaylist(
			env,
			undefined,
			(env: DefaultEnvironment, playlistId: RundownPlaylistId, rundownId: RundownId) => {
				const rundown: DBRundown = {
					peripheralDeviceId: env.ingestDevice._id,
					organizationId: null,
					studioId: env.studio._id,
					showStyleBaseId: env.showStyleBase._id,
					showStyleVariantId: env.showStyleVariant._id,
					playlistId: playlistId,
					_rank: 0,
					_id: rundownId,
					externalId: 'MOCK_RUNDOWN',
					name: 'Default Rundown',
					created: getCurrentTime(),
					modified: getCurrentTime(),
					importVersions: {
						studio: '',
						showStyleBase: '',
						showStyleVariant: '',
						blueprint: '',
						core: '',
					},
					externalNRCSName: 'mock',
					timing: {
						type: PlaylistTimingType.None,
					},
				}
				Rundowns.insert(rundown)

				function createMockPart(index: number, segId: SegmentId): DBPart {
					return {
						_id: protectString(rundownId + '_part' + index),
						segmentId: segId,
						rundownId: rundown._id,
						_rank: index,
						externalId: 'MOCK_PART_' + index,
						title: 'Part ' + index,
					}
				}

				const segmentId0: SegmentId = getRandomId()
				const segmentId1: SegmentId = getRandomId()
				const segmentId2: SegmentId = getRandomId()

				partIds = [
					Parts.insert(createMockPart(0, segmentId0)),
					Parts.insert(createMockPart(1, segmentId0)),
					Parts.insert(createMockPart(2, segmentId0)),
					Parts.insert(createMockPart(3, segmentId0)),
					Parts.insert(createMockPart(4, segmentId0)),

					Parts.insert(createMockPart(10, segmentId1)),
					Parts.insert(createMockPart(11, segmentId1)),
					Parts.insert(createMockPart(12, segmentId1)),

					Parts.insert(createMockPart(20, segmentId2)),
					Parts.insert(createMockPart(21, segmentId2)),
					Parts.insert(createMockPart(22, segmentId2)),
				]

				return rundownId
			}
		))
	})

	function expectLookaheadForLayerMock(
		playlistId0: RundownPlaylistId,
		partInstances: PartInstanceAndPieceInstances[],
		previous: PartInstanceAndPieceInstances | undefined,
		orderedPartsFollowingPlayhead: PartAndPieces[]
	) {
		const playlist = RundownPlaylists.findOne(playlistId0) as RundownPlaylist
		expect(playlist).toBeTruthy()

		expect(findLookaheadForLayerMock).toHaveBeenCalledTimes(2)
		expect(findLookaheadForLayerMock).toHaveBeenNthCalledWith(
			1,
			playlist.currentPartInstanceId,
			partInstances,
			previous,
			orderedPartsFollowingPlayhead,
			'PRELOAD',
			1,
			LOOKAHEAD_DEFAULT_SEARCH_DISTANCE
		)
		expect(findLookaheadForLayerMock).toHaveBeenNthCalledWith(
			2,
			playlist.currentPartInstanceId,
			partInstances,
			previous,
			orderedPartsFollowingPlayhead,
			'WHEN_CLEAR',
			1,
			LOOKAHEAD_DEFAULT_SEARCH_DISTANCE
		)
		findLookaheadForLayerMock.mockClear()
	}

	testInFiber('No pieces', () => {
		const playlist = RundownPlaylists.findOne(playlistId) as RundownPlaylist
		expect(playlist).toBeTruthy()

		const partInstancesInfo: SelectedPartInstancesTimelineInfo = {}

		const fakeParts = partIds.map((p) => ({ part: new Part({ _id: p } as any), pieces: [] }))
		getOrderedPartsAfterPlayheadMock.mockReturnValueOnce(fakeParts.map((p) => p.part))

		const res = runPlayoutOperationWithCache(
			null,
			'test',
			playlistId,
			PlayoutLockFunctionPriority.USER_PLAYOUT,
			null,
			(cache) => getLookeaheadObjects(cache, partInstancesInfo)
		)
		expect(res).toHaveLength(0)

		expect(getOrderedPartsAfterPlayheadMock).toHaveBeenCalledTimes(1)
		expect(getOrderedPartsAfterPlayheadMock).toHaveBeenCalledWith(expect.anything(), 10) // default distance
		expectLookaheadForLayerMock(playlistId, [], undefined, fakeParts)
	})

	function fakeResultObj(id: string, pieceId: string, layer: string): TimelineObjRundown & OnGenerateTimelineObjExt {
		return {
			id: id,
			pieceInstanceId: pieceId,
			layer: layer,
			content: { id },
		} as any
	}

	testInFiber('got some objects', () => {
		const partInstancesInfo: SelectedPartInstancesTimelineInfo = {}

		const fakeParts = partIds.map((p) => ({ part: new Part({ _id: p } as any), pieces: [] }))
		getOrderedPartsAfterPlayheadMock.mockReturnValueOnce(fakeParts.map((p) => p.part))

		findLookaheadForLayerMock
			.mockImplementationOnce((_id, _parts, _prev, _parts2, layer) => ({
				timed: [fakeResultObj('obj0', 'piece0', layer), fakeResultObj('obj1', 'piece1', layer)],
				future: [
					fakeResultObj('obj2', 'piece0', layer),
					fakeResultObj('obj3', 'piece0', layer),
					fakeResultObj('obj4', 'piece0', layer),
				],
			}))
			.mockImplementationOnce((_id, _parts, _prev, _parts2, layer) => ({
				timed: [fakeResultObj('obj5', 'piece1', layer), fakeResultObj('obj6', 'piece0', layer)],
				future: [
					fakeResultObj('obj7', 'piece1', layer),
					fakeResultObj('obj8', 'piece1', layer),
					fakeResultObj('obj9', 'piece0', layer),
				],
			}))

		const res = runPlayoutOperationWithCache(
			null,
			'test',
			playlistId,
			PlayoutLockFunctionPriority.USER_PLAYOUT,
			null,
			(cache) => getLookeaheadObjects(cache, partInstancesInfo)
		)
		expect(res).toMatchSnapshot()

		expect(getOrderedPartsAfterPlayheadMock).toHaveBeenCalledTimes(1)
		expect(getOrderedPartsAfterPlayheadMock).toHaveBeenCalledWith(expect.anything(), 10) // default distance
		expectLookaheadForLayerMock(playlistId, [], undefined, fakeParts)
	})

	testInFiber('Different max distances', () => {
		const partInstancesInfo: SelectedPartInstancesTimelineInfo = {}

		// Set really low
		env.studio.mappings['WHEN_CLEAR'].lookaheadMaxSearchDistance = 0
		env.studio.mappings['PRELOAD'].lookaheadMaxSearchDistance = 0
		runPlayoutOperationWithCache(
			null,
			'test',
			playlistId,
			PlayoutLockFunctionPriority.USER_PLAYOUT,
			null,
			(cache) => getLookeaheadObjects(cache, partInstancesInfo)
		)
		expect(getOrderedPartsAfterPlayheadMock).toHaveBeenCalledTimes(1)
		expect(getOrderedPartsAfterPlayheadMock).toHaveBeenCalledWith(expect.anything(), 0)

		// really high
		getOrderedPartsAfterPlayheadMock.mockClear()
		env.studio.mappings['WHEN_CLEAR'].lookaheadMaxSearchDistance = -1
		env.studio.mappings['PRELOAD'].lookaheadMaxSearchDistance = 2000
		runPlayoutOperationWithCache(
			null,
			'test',
			playlistId,
			PlayoutLockFunctionPriority.USER_PLAYOUT,
			null,
			(cache) => getLookeaheadObjects(cache, partInstancesInfo)
		)
		expect(getOrderedPartsAfterPlayheadMock).toHaveBeenCalledTimes(1)
		expect(getOrderedPartsAfterPlayheadMock).toHaveBeenCalledWith(expect.anything(), 2000)

		// unset
		getOrderedPartsAfterPlayheadMock.mockClear()
		env.studio.mappings['WHEN_CLEAR'].lookaheadMaxSearchDistance = undefined
		env.studio.mappings['PRELOAD'].lookaheadMaxSearchDistance = -1
		runPlayoutOperationWithCache(
			null,
			'test',
			playlistId,
			PlayoutLockFunctionPriority.USER_PLAYOUT,
			null,
			(cache) => getLookeaheadObjects(cache, partInstancesInfo)
		)
		expect(getOrderedPartsAfterPlayheadMock).toHaveBeenCalledTimes(1)
		expect(getOrderedPartsAfterPlayheadMock).toHaveBeenCalledWith(expect.anything(), 10)
	})

	testInFiber('PartInstances translation', () => {
		const fakeParts = partIds.map((p) => ({ part: new Part({ _id: p } as any), pieces: [] }))
		getOrderedPartsAfterPlayheadMock.mockReturnValue(fakeParts.map((p) => p.part))

		const partInstancesInfo: SelectedPartInstancesTimelineInfo = {
			previous: {
				partInstance: 'abc' as any,
				nowInPart: 987,
				pieceInstances: ['1', '2'] as any,
			},
		}
		const expectedPrevious = {
			part: partInstancesInfo.previous!.partInstance,
			onTimeline: true,
			nowInPart: partInstancesInfo.previous!.nowInPart,
			allPieces: partInstancesInfo.previous!.pieceInstances,
		}

		// With a previous
		runPlayoutOperationWithCache(
			null,
			'test',
			playlistId,
			PlayoutLockFunctionPriority.USER_PLAYOUT,
			null,
			(cache) => getLookeaheadObjects(cache, partInstancesInfo)
		)
		expectLookaheadForLayerMock(playlistId, [], expectedPrevious, fakeParts)

		// Add a current
		partInstancesInfo.current = {
			partInstance: { _id: 'curr', part: {} } as any,
			nowInPart: 56,
			pieceInstances: ['3', '4'] as any,
		}
		const expectedCurrent = {
			part: partInstancesInfo.current!.partInstance,
			onTimeline: true,
			nowInPart: partInstancesInfo.current!.nowInPart,
			allPieces: partInstancesInfo.current!.pieceInstances,
		}
		runPlayoutOperationWithCache(
			null,
			'test',
			playlistId,
			PlayoutLockFunctionPriority.USER_PLAYOUT,
			null,
			(cache) => getLookeaheadObjects(cache, partInstancesInfo)
		)
		expectLookaheadForLayerMock(playlistId, [expectedCurrent], expectedPrevious, fakeParts)

		// Add a next
		partInstancesInfo.next = {
			partInstance: 'nxt' as any,
			nowInPart: -85,
			pieceInstances: ['5'] as any,
		}
		const expectedNext = {
			part: partInstancesInfo.next!.partInstance,
			onTimeline: false,
			nowInPart: partInstancesInfo.next!.nowInPart,
			allPieces: partInstancesInfo.next!.pieceInstances,
		}
		runPlayoutOperationWithCache(
			null,
			'test',
			playlistId,
			PlayoutLockFunctionPriority.USER_PLAYOUT,
			null,
			(cache) => getLookeaheadObjects(cache, partInstancesInfo)
		)
		expectLookaheadForLayerMock(playlistId, [expectedCurrent, expectedNext], expectedPrevious, fakeParts)

		// current has autonext
		partInstancesInfo.current.partInstance.part.autoNext = true
		expectedNext.onTimeline = true
		runPlayoutOperationWithCache(
			null,
			'test',
			playlistId,
			PlayoutLockFunctionPriority.USER_PLAYOUT,
			null,
			(cache) => getLookeaheadObjects(cache, partInstancesInfo)
		)
		expectLookaheadForLayerMock(playlistId, [expectedCurrent, expectedNext], expectedPrevious, fakeParts)
	})

	// testInFiber('Pieces', () => {
	// 	const fakeParts = partIds.map((p) => ({ _id: p })) as Part[]
	// 	getOrderedPartsAfterPlayheadMock.mockReturnValue(fakeParts)

	// 	const partInstancesInfo: SelectedPartInstancesTimelineInfo = {}

	// 	const pieceMap = new Map<PartId, Piece[]>()

	// 	partIds.forEach((id, i) => {
	// 		const pieces: Piece[] = []
	// 		const target = (i + 2) % 3
	// 		for (let o = 0; o <= target; o++) {
	// 			const piece: Piece = {
	// 				_id: `piece_${i}_${o}`,
	// 				startPartId: id,
	// 				startRundownId: rundownId,
	// 				invalid: target === 2 && i % 2 === 1, // Some 'randomness'
	// 			} as any
	// 			Pieces.insert(piece)
	// 			if (!piece.invalid) {
	// 				pieces.push(piece)
	// 			}
	// 		}

	// 		if (pieces.length > 0) {
	// 			pieceMap.set(id, pieces)
	// 		}
	// 	})
	// 	// Add a couple of pieces which are incorrectly owned
	// 	Pieces.insert({
	// 		_id: `piece_a`,
	// 		startPartId: partIds[0],
	// 		startRundownId: rundownId + '9',
	// 	} as any)
	// 	Pieces.insert({
	// 		_id: `piece_b`,
	// 		startPartId: 'not-real',
	// 		startRundownId: rundownId,
	// 	} as any)

	// 	// pieceMap should have come through valid
	// 	;(
	// 		wrapWithCacheForRundownPlaylist(playlist, async (cache) => {
	// 			await getLookeaheadObjects(cache, env.studio, playlist, partInstancesInfo)
	// 			expect(cache.Pieces.initialized).toBeFalsy()
	// 		})
	// 	)
	// 	expectLookaheadForLayerMock(playlist, [], undefined, fakeParts, pieceMap)

	// 	// Use the modified cache values
	// 	const removedIds: PieceId[] = protectStringArray(['piece_1_0', 'piece_4_0'])
	// 	;(
	// 		wrapWithCacheForRundownPlaylist(playlist, async (cache) => {
	// 			expect(
	// 				cache.Pieces.update(removedIds[0], {
	// 					$set: {
	// 						invalid: true,
	// 					},
	// 				})
	// 			).toEqual(1)
	// 			cache.Pieces.remove(removedIds[1])
	// 			expect(cache.Pieces.initialized).toBeTruthy()

	// 			await getLookeaheadObjects(cache, env.studio, playlist, partInstancesInfo)
	// 		})
	// 	)
	// 	const pieceMap2 = new Map<PartId, Piece[]>()
	// 	for (const [id, pieces] of pieceMap) {
	// 		const pieces2 = pieces.filter((p) => !removedIds.includes(p._id))
	// 		if (pieces2.length) pieceMap2.set(id, pieces2)
	// 	}
	// 	expectLookaheadForLayerMock(playlist, [], undefined, fakeParts, pieceMap2)
	// })
})
