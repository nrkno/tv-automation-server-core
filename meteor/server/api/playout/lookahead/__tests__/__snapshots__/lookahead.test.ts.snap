// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lookahead got some objects 1`] = `
Array [
  Object {
    "content": Object {
      "id": "obj0",
    },
    "disabled": false,
    "enable": Object {
      "end": "#piece0obj0.start",
      "start": 1,
    },
    "id": "lookahead_timed0_obj0",
    "isLookahead": true,
    "layer": "PRELOAD_lookahead",
    "lookaheadForLayer": "PRELOAD",
    "pieceInstanceId": "piece0",
    "priority": 0.1,
  },
  Object {
    "content": Object {
      "id": "obj1",
    },
    "disabled": false,
    "enable": Object {
      "end": "#piece1obj1.start",
      "start": "#piece0obj0.start + 0",
    },
    "id": "lookahead_timed1_obj1",
    "isLookahead": true,
    "layer": "PRELOAD_lookahead",
    "lookaheadForLayer": "PRELOAD",
    "pieceInstanceId": "piece1",
    "priority": 0.1,
  },
  Object {
    "content": Object {
      "id": "obj2",
    },
    "disabled": false,
    "enable": Object {
      "while": "1",
    },
    "id": "lookahead_future0_obj2",
    "isLookahead": true,
    "layer": "PRELOAD_lookahead",
    "lookaheadForLayer": "PRELOAD",
    "pieceInstanceId": "piece0",
    "priority": 0.07500000000000001,
  },
  Object {
    "content": Object {
      "id": "obj3",
    },
    "disabled": false,
    "enable": Object {
      "while": "1",
    },
    "id": "lookahead_future1_obj3",
    "isLookahead": true,
    "layer": "PRELOAD_lookahead",
    "lookaheadForLayer": "PRELOAD",
    "pieceInstanceId": "piece0",
    "priority": 0.05,
  },
  Object {
    "content": Object {
      "id": "obj4",
    },
    "disabled": false,
    "enable": Object {
      "while": "1",
    },
    "id": "lookahead_future2_obj4",
    "isLookahead": true,
    "layer": "PRELOAD_lookahead",
    "lookaheadForLayer": "PRELOAD",
    "pieceInstanceId": "piece0",
    "priority": 0.025,
  },
  Object {
    "content": Object {
      "id": "obj5",
    },
    "disabled": false,
    "enable": Object {
      "end": "#piece1obj5.start",
      "start": 1,
    },
    "id": "lookahead_timed0_obj5",
    "isLookahead": true,
    "layer": "WHEN_CLEAR",
    "pieceInstanceId": "piece1",
    "priority": 0.1,
  },
  Object {
    "content": Object {
      "id": "obj6",
    },
    "disabled": false,
    "enable": Object {
      "end": "#piece0obj6.start",
      "start": "#piece1obj5.start + 0",
    },
    "id": "lookahead_timed1_obj6",
    "isLookahead": true,
    "layer": "WHEN_CLEAR",
    "pieceInstanceId": "piece0",
    "priority": 0.1,
  },
  Object {
    "content": Object {
      "id": "obj7",
    },
    "disabled": false,
    "enable": Object {
      "start": "#piece0obj6.start + 0",
    },
    "id": "lookahead_future0_obj7",
    "isLookahead": true,
    "layer": "WHEN_CLEAR",
    "pieceInstanceId": "piece1",
    "priority": 0.07500000000000001,
  },
  Object {
    "content": Object {
      "id": "obj8",
    },
    "disabled": true,
    "enable": Object {
      "while": "1",
    },
    "id": "lookahead_future1_obj8",
    "isLookahead": true,
    "layer": "WHEN_CLEAR",
    "pieceInstanceId": "piece1",
    "priority": 0.05,
  },
  Object {
    "content": Object {
      "id": "obj9",
    },
    "disabled": true,
    "enable": Object {
      "while": "1",
    },
    "id": "lookahead_future2_obj9",
    "isLookahead": true,
    "layer": "WHEN_CLEAR",
    "pieceInstanceId": "piece0",
    "priority": 0.025,
  },
]
`;
